Providing a basic documentation (#27)
MatterMiners/lapis: docs/source/topics/concept.rst

Simulation Concept
==================

Background
----------

.. TODO::

    HEP context.

Components
----------

The core simulation builds on several components, and concepts:

* :term:`Job Generator`,
* :term:`Job Queue`,
* :term:`Pools <Pool>` and their :term:`Controllers <Controller>`,
* :term:`Drones <Drone>`, and
* the :term:`Scheduler`,

If you are planning to adapt the simulation for your specific use case, please
consider the different components to determine what and where to extend functionality.

Job Generator
~~~~~~~~~~~~~

> The Job Generator works on the job input files. It takes care to

Reviewer: ```suggestion
The Job Generator processes any job input files. It takes care to
```

into the simulations :term:`Job Queue` as soon as the simulation time corresponds
to the translated :term:`job` queueing time.

Job Queue
~~~~~~~~~

The Job Queue is filled with :term:`jobs <Job>` in creation-time order by the
:term:`Job Generator`. The queue is managed by the :term:`scheduler` and contains
all :term:`jobs <Job>` that are not yet scheduled to a :term:`drone` as well as
:term:`jobs <Job>` that have not yet been processed succesfully.

Pools
~~~~~

Pools are created based on the pool input files. Each pool is characterised by
a set of defined resources. Further, pools have a ``capacity`` number of
:term:`drones <Drone>` that can be created from a given pool. If the capacity
is not specified, a maximum capacity of ``float("inf")`` is assumed.

For pools, we differentiate static and dynamic pools. While static pools are
intialised with a fixed amount of :term:`drones <Drone>`, the number of
:term:`drones <Drone>` is adapted dynamically by the
:term:`pool controller <Controller>` for dynamic pools.

.. autoclass:: lapis.pool.Pool
.. autoclass:: lapis.pool.StaticPool

Controllers
~~~~~~~~~~~

> Each :term:`pool` is started with an appropriate controller. Each controller runs

Reviewer: ```suggestion
Each :term:`pool` is managed by a parent controller. Each controller runs
```

Author: I wouldn't like to put the information about parent here... This is a piece of information that we do not cover. What do you think about:

```suggestion
Each :term:`pool` is managed by a controller. Each controller runs
```

of future extensions.

Available controller implementations from COBalD in LAPIS are:

.. autoclass:: lapis.controller.SimulatedLinearController
    :members:

.. autoclass:: lapis.controller.SimulatedRelativeSupplyController
    :members:

And there is also an implementation considered as an extension for COBalD:

.. autoclass:: lapis.controller.SimulatedCostController
    :members:

Drones
~~~~~~

Drones provide instances of the set of resources defined by a given :term:`pool`.
Drones are the only objects in the simulation that are able to process
:term:`jobs <Job>`. Simplified, drones represent worker nodes.

The concept of drones is introduced by TARDIS. A drone is a generalisation of
the pilot concept used for example in High Energy Physics and is a placeholder
for the real workloads to be processed. A drone is expected to autonomously
manage its lifecycle, meaning, that it handles failures and termination
independently from other components within the system.

.. warning::

    Drones are not yet fully employed in LAPIS. They already run independently
    but do not handle termination themselves.

Scheduler
~~~~~~~~~

> The scheduler is the connecting component between the :term:`jobs <Job>` in the

Reviewer: This section is very technically detailed. Perhaps a more general description matching other scheduler implementations would be more appropriate.

Author: What do you mean? Could you clarify a bit?
My intention is to follow the descriptions before. Those followed the actual implementation within LAPIS, so this is what I also did here for the scheduler.
